File: sdk/examples/client/output/build_nft_output.rs
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! In this example we will build an NFT output.
//!
//! Rename `.env.example` to `.env` first, then run the command:
//! ```sh
//! cargo run --release --example build_nft_output [ADDRESS]
//! ```

use iota_sdk::{
    client::{Client, Result},
    types::block::{
        address::Address,
        output::{
            feature::{IssuerFeature, MetadataFeature, SenderFeature, TagFeature},
            unlock_condition::AddressUnlockCondition,
            NftId, NftOutputBuilder,
        },
    },
};

const MUTABLE_METADATA: &str = "mutable metadata";
const TAG: &str = "my tag";

#[tokio::main]
async fn main() -> Result<()> {
    // This example uses secrets in environment variables for simplicity which should not be done in production.
    dotenvy::dotenv().ok();

    for var in ["NODE_URL"] {
        std::env::var(var).unwrap_or_else(|_| panic!(".env variable '{var}' is undefined, see .env.example"));
    }

    // Create a node client.
    let client = Client::builder()
        .with_node(&std::env::var("NODE_URL").unwrap())?
        .finish()
        .await?;

    let token_supply = client.get_token_supply().await?;
    let rent_structure = client.get_rent_structure().await?;

    let address = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "rms1qpllaj0pyveqfkwxmnngz2c488hfdtmfrj3wfkgxtk4gtyrax0jaxzt70zy".to_string());
    let address = Address::try_from_bech32(address)?;

    // IOTA NFT Standard - IRC27: https://github.com/iotaledger/tips/blob/main/tips/TIP-0027/tip-0027.md
    let tip_27_immutable_metadata = serde_json::from_str::<serde_json::Value>(
        r#"
        {
            "standard": "IRC27",
            "version": "v1.0",
            "type":"image/jpeg",
            "uri":"https://mywebsite.com/my-nft-files-1.jpeg",
            "name":"My NFT #0001"
        }"#,
    )?
    .to_string();

    // NftId needs to be null the first time
    let nft_output = NftOutputBuilder::new_with_minimum_storage_deposit(rent_structure, NftId::null())
        .add_unlock_condition(AddressUnlockCondition::new(address))
        .add_feature(SenderFeature::new(address))
        .add_feature(MetadataFeature::new(MUTABLE_METADATA)?)
        .add_feature(TagFeature::new(TAG)?)
        .add_immutable_feature(IssuerFeature::new(address))
        .add_immutable_feature(MetadataFeature::new(tip_27_immutable_metadata)?)
        .finish_output(token_supply)?;

    println!("{nft_output:#?}");

    Ok(())
}

File: sdk/src/client/bindings/java/examples/src/BuildNftOutput.java
import org.iota.Client;
import org.iota.types.ClientConfig;
import org.iota.types.exceptions.ClientException;
import org.iota.types.Output;
import org.iota.types.UnlockCondition;
import org.iota.types.exceptions.InitializeClientException;
import org.iota.types.ids.NftId;
import org.iota.types.output_builder.NftOutputBuilderParams;
import org.iota.types.secret.GenerateAddressesOptions;
import org.iota.types.secret.MnemonicSecretManager;
import org.iota.types.secret.Range;
import org.iota.types.Feature;

import com.google.gson.*;
import org.apache.commons.codec.binary.Hex;

public class BuildNftOutput {
  public static void main(String[] args) throws ClientException, InitializeClientException {
    // Build the client.
    Client client = new Client(new ClientConfig().withNodes(new String[] {
      "https://api.testnet.shimmer.network"
    }));

    // Configure an NFT output.
    String hexAddress = client.bech32ToHex("rms1qpllaj0pyveqfkwxmnngz2c488hfdtmfrj3wfkgxtk4gtyrax0jaxzt70zy");
    // NftId needs to be null the first time
    NftId nftId = new NftId("0x0000000000000000000000000000000000000000000000000000000000000000");
    UnlockCondition[] unlockConditions = new UnlockCondition[] {
      new UnlockCondition("{ type: 0, address: { type: 0, pubKeyHash: \"" + hexAddress + "\" } }")
    };

    // IOTA NFT Standard - IRC27: https://github.com/iotaledger/tips/blob/main/tips/TIP-0027/tip-0027.md
    JsonObject tip27ImmutableMetadata = new JsonObject();
    tip27ImmutableMetadata.addProperty("standard", "IRC27");
    tip27ImmutableMetadata.addProperty("version", "v1.0");
    tip27ImmutableMetadata.addProperty("type", "image/jpeg");
    tip27ImmutableMetadata.addProperty("uri", "https://mywebsite.com/my-nft-files-1.jpeg");
    tip27ImmutableMetadata.addProperty("name", "My NFT #0001");

    Feature[] immutableFeatures = new Feature[] {
      // issuer feature
      new Feature("{ type: 1, address: { type: 0, pubKeyHash: \"" + hexAddress + "\" } }"),
        // metadata feature
        new Feature("{ type: 2, data: \"0x" + Hex.encodeHexString(tip27ImmutableMetadata.toString().getBytes()) + "\" }")
    };
    Feature[] features = new Feature[] {
      // sender feature
      new Feature("{ type: 0, address: { type: 0, pubKeyHash: \"" + hexAddress + "\" } }"),
        // metadata feature
        new Feature("{ type: 2, data: \"0x" + Hex.encodeHexString("mutable metadata".getBytes()) + "\" }"),
        // tag feature
        new Feature("{ type: 3, tag: \"0x" + Hex.encodeHexString("my tag".getBytes()) + "\" }")
    };

    NftOutputBuilderParams params = new NftOutputBuilderParams()
      .withNftId(nftId)
      .withUnlockConditions(unlockConditions)
      .withImmutableFeatures(immutableFeatures)
      .withFeatures(features);

    // Build the output.
    Output output = client.buildNftOutput(params);

    // Print the output.
    System.out.println(
        new GsonBuilder().setPrettyPrinting().create().toJson(JsonParser.parseString(output.toString()))
    );
  }
}